devices.go

```diff
@@ -331,26 +331,21 @@ type DeviceServiceOp struct {
 // Device properties: Hostname, Description, Tags, ID, ShortID, Network.Address,
 // Plan.Name, Plan.Slug, Facility.Code, Facility.Name, OS.Name, OS.Slug,
 // HardwareReservation.ID, HardwareReservation.ShortID
-func (s *DeviceServiceOp) List(projectID string, listOpt *ListOptions) (devices []Device, resp *Response, err error) {
-	listOpt = listOpt.Including("facility")
-	params := urlQuery(listOpt)
-	path := fmt.Sprintf("%s/%s%s?%s", projectBasePath, projectID, deviceBasePath, params)
+func (s *DeviceServiceOp) List(projectID string, opts *ListOptions) (devices []Device, resp *Response, err error) {
+	opts = opts.Including("facility")
+	apiCallUrl := getApiCallUrl(opts, projectBasePath, projectID, deviceBasePath)
```

Comment:
@displague what do you think about this approach to build the API call URLs?

Reply:
It seems like a convenient short-hand, but there is a downside to this. We won't have compile-time checks when the URL format string does not match the parameter listed.

My argument here is weak because we don't have a central body of URL paths and formats (constants representing each URL).

Here's an alternative to consider (and hopefully draw a better pattern from).

```
s.projectURL(opts, projectID)
s.projectURLWithIdentifier(opts, projectID, deviceID, action)
s.projectURLWithIdentifierAction(opts, projectID, deviceID, "action")
```

I'm not sold on this either, especially because we would have to copy this pattern to each service.

There are other approaches we could take that might minimize the amount of copying, possibly with an internal url helper package that defines the URL constants.

Do you have any thoughts on this @detiber?

Reply:
@t0mk maybe we could open this last URL building change in a new PR. I'd like see the user-visible change introduced first and then we can optimize the internals.  I don't see a great deal of benefit in the change and I think it introduces risk (and I don't know if we can come up with a good alternative that we can agree on without delaying this PR).

Were we ready to remove the WIP tag?

Reply:
@displague 

> s.projectURL(opts, projectID)
> s.projectURLWithIdentifier(opts, projectID, deviceID)
> s.projectURLWithIdentifierAction(opts, projectID, deviceID, "action")

I don't think this would make the code look very nice. I think there's not just 3 formats of the API endpoints in the Eq Metal API, so the API Call methods would not look very consistent. IMO just one general variadic function is the better looking alternativ.

But OK, I will not do the GetApiUrl in this PR.

I will get devices and volumes to the state of using the new general struct for options, and then ask you for review again. Once you ack that, I will reformat the rest of Packngo to using the new opt struct (wihout GetApiUrl).